#include "mainwindow.h"
#include "blackboxinterface.h"
#include "emergencycontroller.h"
#include "environmentalcontroller.h"
#include "hardwareinterface.h"
#include "mcp3008interface.h"
#include "pythonaimanager.h"
#include "securitycontroller.h"
#include "ui_mainwindow.h"
#include <QApplication>
#include <QDateTime>
#include <QDebug>
#include <QFile>
#include <QJsonDocument>
#include <QJsonObject>
#include <QKeyEvent>
#include <QMap>
#include <QMessageBox>
#include <QTextStream>

MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent), ui(new Ui::MainWindow) {
  ui->setupUi(this);

  // 1. åˆå§‹åŒ–åº•å±¤ç¡¬é«”ä»‹é¢ (ç›´æ¥ç”±ä¸»åŸ·è¡Œç·’æˆ–é‚è¼¯åŸ·è¡Œç·’ç®¡ç†)
  blackbox = new BlackboxInterface(this);
  adc = new Mcp3008Interface(this);

  // 2. åˆå§‹åŒ–æ¥­å‹™é‚è¼¯æ§åˆ¶å™¨
  camera = new PythonAiManager();
  security = new SecurityController();
  env = new EnvironmentalController();
  emergency = new EmergencyController(blackbox);

  // åˆå§‹åŒ–åˆ—è¡¨æ¨¡å‹ (ç”¨æ–¼é¡¯ç¤ºé»‘ç›’å­äº‹ä»¶)
  eventModel = new QStringListModel(this);
  ui->eventTable->setModel(eventModel);

  // 3. åŸ·è¡Œç·’ç®¡ç†
  cameraThread = new QThread(this);
  logicThread = new QThread(this);

  security->moveToThread(logicThread);
  env->moveToThread(logicThread);

  // 4. é€£ç·šè¨­å®š (è¨Šè™Ÿå‚³é)

  // Camera -> UI (å¾ Python ç²å–å½±åƒèˆ‡ AI çµæœ)
  connect(camera, &PythonAiManager::frameReady, this, &MainWindow::updateFrame);
  connect(camera, &PythonAiManager::detectionAlert, this,
          [this](QString type, double conf) {
            // æª¢æŸ¥å†·å»æ™‚é–“ï¼Œé¿å…æ´—æ¿
            QDateTime now = QDateTime::currentDateTime();
            int cooldown = 60; // é è¨­ 60 ç§’å†·å»

            if (type == "owner")
              cooldown = 10; // ä¸»äºº 10 ç§’å†·å»

            if (m_lastAlertTime.contains(type) &&
                m_lastAlertTime[type].secsTo(now) < cooldown) {
              return; // é‚„åœ¨å†·å»ä¸­ï¼Œä¸è§¸ç™¼
            }

            if (type == "pig") {
              // è±¬è±¬ç‰¹åˆ¥è™•ç†ï¼šå¦‚æœç‚¸å½ˆå·²ç¶“å•Ÿå‹•ï¼Œå°±ä¸å†è§¸ç™¼
              if (emergency->isBombActive())
                return;
              m_lastAlertTime[type] = now;
              simulateAiTrigger("pig");
            } else if (type == "stranger") {
              m_lastAlertTime[type] = now;
              simulateAiTrigger("stranger");
            } else if (type == "owner") {
              m_lastAlertTime[type] = now;
              // ä¸»äººé©—è­‰æˆåŠŸé‚è¼¯
              ui->status_label->setText(
                  QString::fromUtf8("ç‹€æ…‹: æ­¡è¿ä¸»äººå›å®¶ï¼"));
              handleShortcut(1); // åŸ·è¡Œé–‹é–€å‹•ä½œ (äº®ç¶ ç‡ˆ)
            }
          });

  // Security Logic -> Hardware/Log
  connect(security, &SecurityController::requestLog, blackbox,
          &BlackboxInterface::logEvent);
  connect(security, &SecurityController::requestGpio, blackbox,
          &BlackboxInterface::setGpio);
  connect(security, &SecurityController::passwordVerified, this,
          [this](bool success) {
            if (success) {
              ui->status_label->setText(QString::fromUtf8("ç‹€æ…‹: é©—è­‰æˆåŠŸï¼"));

              // æª¢æŸ¥ç›®å‰æ˜¯å¦ç‚ºè±¬è±¬è­¦å ± (ç‚¸å½ˆå•Ÿå‹•ä¸­)
              bool wasPigAlarm = emergency->isBombActive();

              // 1. é€šç”¨è§£é™¤å‹•ä½œ
              blackbox->setGpio(LED_RED, 0);
              blackbox->setGpio(LED_BLUE, 0);
              blackbox->setGpio(BUZZER, 0);

              // 2. é‡å°ä¸åŒè­¦å ±é¡å‹çš„å¾ŒçºŒè™•ç†
              if (wasPigAlarm) {
                // è±¬è±¬è­¦å ±ï¼šè§£é™¤ç‚¸å½ˆ
                emergency->disarmBomb();
              } else {
                // é™Œç”Ÿäººè­¦å ±æˆ–å…¶ä»–ï¼šåŸ·è¡Œé–‹é–€å‹•ä½œ
                handleShortcut(1);
              }

              // æ¸…ç†ç‹€æ…‹æª”æ¡ˆ
              QFile::remove("/tmp/guardian_alarm_status.json");
              QFile::remove("/tmp/guardian_unlock_status.json");

              // åŒæ™‚è§£é™¤é‚è¼¯é–å®š
              QMetaObject::invokeMethod(security, "setAlarmActive",
                                        Q_ARG(bool, false));
            } else {
              ui->status_label->setText(QString::fromUtf8("ç‹€æ…‹: é©—è­‰å¤±æ•—ï¼"));
            }
          });

  // Environmental Logic -> Hardware/UI
  connect(env, &EnvironmentalController::requestGpio, blackbox,
          &BlackboxInterface::setGpio);
  connect(
      env, &EnvironmentalController::lightLevelChanged, this,
      [this](int value, const QString &mode) {
        ui->status_label->setText(
            QString("ç‹€æ…‹: ç³»çµ±é‹ä½œä¸­ | äº®åº¦: %1 (%2)").arg(value).arg(mode));
      });

  // Emergency Logic -> UI
  connect(emergency, &EmergencyController::countdownUpdated, this,
          [this](int totalSeconds, QString formattedTime) {
            ui->status_label->setText(
                QString("<font color='red'>ğŸš¨ ç·Šæ€¥å€’æ•¸: %1 ğŸš¨</font>")
                    .arg(formattedTime));

            // åŒæ­¥æ›´æ–° JSON æª”æ¡ˆçµ¦ Web Server
            updateAlarmJsonWithCountdown(totalSeconds, formattedTime);

            // å€’è¨ˆæ™‚èœ‚é³´å™¨é‚è¼¯ï¼šæ¯ç§’éŸ¿ä¸€ä¸‹ (200ms)
            if (!m_isMuted) {
              blackbox->setGpio(BUZZER, 1);
              QTimer::singleShot(200, [this]() {
                // åªæœ‰åœ¨ç‚¸å½ˆä»åœ¨å•Ÿå‹•ç‹€æ…‹æ™‚æ‰é—œé–‰ï¼Œé¿å…å½±éŸ¿å…¶ä»–é–‹é–€éŸ³æ•ˆ
                if (emergency->isBombActive()) {
                  blackbox->setGpio(BUZZER, 0);
                }
              });
            }
          });

  connect(emergency, &EmergencyController::bombExploded, this, [this]() {
    ui->status_label->setText(
        "<font color='red'><b>ğŸ’¥ ç³»çµ±å·²ç‚¸æ¯€ ğŸ’¥</b></font>");
    QMessageBox::critical(this, "è­¦å‘Š", "å€’æ•¸çµæŸï¼Œç³»çµ±å·²åŸ·è¡Œç·Šæ€¥è‡ªæ¯€ç¨‹åºï¼");
  });

  connect(emergency, &EmergencyController::bombDisarmed, this,
          [this]() { ui->status_label->setText("ç‹€æ…‹: ç·Šæ€¥ç‹€æ…‹å·²è§£é™¤"); });

  // 5. å•Ÿå‹•åŸ·è¡Œç·’èˆ‡æ„Ÿæ¸¬å™¨è¼ªè©¢
  logicThread->start();

  // å•Ÿå‹• Python AI å¼•æ“
  camera->start();

  // æ„Ÿæ¸¬å™¨è¼ªè©¢å®šæ™‚å™¨ (åœ¨ä¸»åŸ·è¡Œç·’ä¸­è§¸ç™¼ï¼Œé€éè¨Šè™Ÿäº¤çµ¦é‚è¼¯åŸ·è¡Œç·’è™•ç†)
  QTimer *sensorTimer = new QTimer(this);
  connect(sensorTimer, &QTimer::timeout, this, &MainWindow::pollSensors);
  sensorTimer->start(1000);

  // åˆå§‹åŒ–æ‰‹å‹• LED ç‹€æ…‹
  m_isAutoLight = true;
  m_manualYellowLed = false;

  setupShortcuts();
  connect(ui->password_input, &QLineEdit::returnPressed, this,
          &MainWindow::handlePasswordInput);
}

MainWindow::~MainWindow() {
  camera->stop();
  // cameraThread ä¸å†ç”¨æ–¼ PythonAiManagerï¼Œä½†è‹¥æœ‰å…¶ä»–ç”¨é€”å¯ä¿ç•™
  if (cameraThread->isRunning()) {
    cameraThread->quit();
    cameraThread->wait();
  }

  logicThread->quit();
  logicThread->wait();

  delete ui;
}

void MainWindow::pollSensors() {
  int lightValue = adc->readAdc(0);
  if (lightValue >= 0) {
    QMetaObject::invokeMethod(env, "updateLightLevel", Q_ARG(int, lightValue));
  }

  // --- æ–°å¢ï¼šè®€å–é»‘ç›’å­æ—¥èªŒä¸¦æ›´æ–° UI åˆ—è¡¨ ---
  QString newLogs = blackbox->readLogs();
  if (!newLogs.isEmpty()) {
    QStringList newLines = newLogs.split('\n', QString::SkipEmptyParts);
    m_logHistory.append(newLines);

    // é™åˆ¶é¡¯ç¤ºç­†æ•¸ï¼Œé¿å…è¨˜æ†¶é«”ä½”ç”¨éå¤§ (ä¾‹å¦‚ä¿ç•™æœ€å¾Œ 100 ç­†)
    while (m_logHistory.size() > 100) {
      m_logHistory.removeFirst();
    }

    eventModel->setStringList(m_logHistory);
    ui->eventTable->scrollToBottom();
  }

  // --- æ–°å¢ï¼šæª¢æŸ¥ Web Server çš„é ç«¯æŒ‡ä»¤ ---
  QFile controlFile("/tmp/guardian_control.txt");
  if (controlFile.exists() &&
      controlFile.open(QIODevice::ReadOnly | QIODevice::Text)) {
    QString action = controlFile.readAll().trimmed();
    controlFile.close();

    qDebug() << "åµæ¸¬åˆ°é ç«¯æŒ‡ä»¤:" << action;

    if (action == "open_door") {
      handleShortcut(1); // åŸ·è¡Œäº®ç¶ ç‡ˆé‚è¼¯
    } else if (action == "mute_alarm") {
      handleShortcut(2); // åŸ·è¡ŒéœéŸ³é‚è¼¯ (åƒ…é—œé–‰èœ‚é³´å™¨èˆ‡ LED)
      // emergency->disarmBomb(); // ç§»é™¤ï¼šé ç«¯éœéŸ³ä¹Ÿä¸æ‡‰è§£é™¤ç‚¸å½ˆå€’æ•¸
    } else if (action == "reset") {
      // é‡ç½®è­¦å ±ç‹€æ…‹æª”æ¡ˆèˆ‡ç·Šæ€¥ç‹€æ…‹
      QFile::remove("/tmp/guardian_alarm_status.json");
      emergency->disarmBomb(); // ç¢ºä¿åœæ­¢ Kernel Driver çš„ç·Šæ€¥è¨ˆæ™‚èˆ‡çˆ†ç‚¸è§¸ç™¼
      ui->status_label->setText(QString::fromUtf8("ç‹€æ…‹: ç³»çµ±å·²é ç«¯é‡ç½®"));
      blackbox->logEvent("ç³»çµ±ç¶“ç”±é ç«¯ç¶²é é‡ç½®", 0);
    } else if (action == "test_alarm") {
      // æ¨¡æ“¬ AI è§¸ç™¼è­¦å ±
      simulateAiTrigger("pig");
    }

    QFile::remove("/tmp/guardian_control.txt"); // åŸ·è¡Œå¾Œåˆªé™¤æŒ‡ä»¤æª”æ¡ˆ
  }

  // --- æ–°å‚³ï¼šæª¢æŸ¥ é ç«¯è§£é–ç‹€æ…‹ (åƒ…åœ¨è­¦å ±å•Ÿå‹•æ™‚æœ‰æ•ˆ) ---
  if (QFile::exists("/tmp/guardian_alarm_status.json")) {
    QFile unlockFile("/tmp/guardian_unlock_status.json");
    if (unlockFile.exists() &&
        unlockFile.open(QIODevice::ReadOnly | QIODevice::Text)) {
      QString data = unlockFile.readAll();
      unlockFile.close();

      // å¦‚æœé ç«¯å·²æˆæ¬Š (remote_unlocked) ä¸”å°šæœªç”Ÿæˆç¾å ´éš¨æ©Ÿç¢¼
      QJsonDocument doc = QJsonDocument::fromJson(data.toUtf8());
      QJsonObject obj = doc.object();

      if (obj["remote_unlocked"].toBool() &&
          !obj["random_code_generated"].toBool()) {

        // æ­¤æ™‚æ‰ç”Ÿæˆéš¨æ©Ÿç¢¼ï¼Œä¸¦ç™¼é€è‡³ Discord
        QString code =
            security->generateRandomCode(false); // ä¸åœ¨æœ¬åœ°é¡¯ç¤ºé©—è­‰ç¢¼
        ui->status_label->setText(
            QString::fromUtf8("é ç«¯æˆæ¬Šé€šéï¼é©—è­‰ç¢¼å·²ç™¼é€è‡³æ‚¨çš„ Discord"));
        sendDiscordCode(code);

        // æ›´æ–°ç‹€æ…‹ï¼Œæ¨™è¨˜å·²ç”Ÿæˆï¼Œé¿å…é‡è¤‡
        obj["random_code_generated"] = true;
        obj["random_code"] = code;

        if (unlockFile.open(QIODevice::WriteOnly | QIODevice::Text)) {
          unlockFile.write(QJsonDocument(obj).toJson());
          unlockFile.close();
        }
      }
    }
  }
}

void MainWindow::setupShortcuts() {
  QShortcut *f1 = new QShortcut(QKeySequence(Qt::Key_F1), this);
  connect(f1, &QShortcut::activated, this, [this]() { handleShortcut(1); });

  QShortcut *f2 = new QShortcut(QKeySequence(Qt::Key_F2), this);
  connect(f2, &QShortcut::activated, this, [this]() { handleShortcut(2); });

  QShortcut *f3 = new QShortcut(QKeySequence(Qt::Key_F3), this);
  connect(f3, &QShortcut::activated, this, [this]() { handleShortcut(3); });

  QShortcut *f4 = new QShortcut(QKeySequence(Qt::Key_F4), this);
  connect(f4, &QShortcut::activated, this, [this]() { handleShortcut(4); });

  // --- æ–°å¢ï¼šF5 æ¨¡æ“¬ AI è§¸ç™¼ ---
  QShortcut *f5 = new QShortcut(QKeySequence(Qt::Key_F5), this);
  connect(f5, &QShortcut::activated, this, [this]() {
    simulateAiTrigger("pig");
    ui->status_label->setText(QString::fromUtf8("ç‹€æ…‹: [F5] æ¨¡æ“¬å°è±¬å…¥ä¾µè­¦å ±"));
  });

  // --- æ–°å¢ï¼šF6 æ¨¡æ“¬ é™Œç”Ÿäºº è§¸ç™¼ ---
  QShortcut *f6 = new QShortcut(QKeySequence(Qt::Key_F6), this);
  connect(f6, &QShortcut::activated, this, [this]() {
    simulateAiTrigger("stranger");
    ui->status_label->setText(QString::fromUtf8("ç‹€æ…‹: [F6] æ¨¡æ“¬é™Œç”Ÿäººåµæ¸¬"));
  });

  // --- æ–°å¢ï¼šF7 åˆ‡æ›è‡ªå‹•/æ‰‹å‹•ç‡ˆå…‰ ---
  QShortcut *f7 = new QShortcut(QKeySequence(Qt::Key_F7), this);
  connect(f7, &QShortcut::activated, this, [this]() {
    m_isAutoLight = !m_isAutoLight;
    QMetaObject::invokeMethod(env, "setAutoMode", Q_ARG(bool, m_isAutoLight));
    QString modeStr = m_isAutoLight ? "è‡ªå‹• (å…‰æ•æ§åˆ¶)" : "æ‰‹å‹• (å¿«æ·éµæ§åˆ¶)";
    ui->status_label->setText(
        QString("ç‹€æ…‹: [F7] ç‡ˆå…‰æ¨¡å¼æ”¹ç‚º %1").arg(modeStr));
    blackbox->logEvent(QString("ç‡ˆå…‰æ¨¡å¼åˆ‡æ›: %1").arg(modeStr), 0);
  });

  // --- æ–°å¢ï¼šF8 æ‰‹å‹•é–‹é—œé»ƒç‡ˆ ---
  QShortcut *f8 = new QShortcut(QKeySequence(Qt::Key_F8), this);
  connect(f8, &QShortcut::activated, this, [this]() {
    if (m_isAutoLight) {
      ui->status_label->setText(
          "ç‹€æ…‹: [F8] ç›®å‰ç‚ºè‡ªå‹•æ¨¡å¼ï¼Œè«‹å…ˆæŒ‰ F7 åˆ‡æ›è‡³æ‰‹å‹•");
    } else {
      m_manualYellowLed = !m_manualYellowLed;
      QMetaObject::invokeMethod(env, "setManualLed",
                                Q_ARG(bool, m_manualYellowLed));
      QString stateStr = m_manualYellowLed ? "é–‹å•Ÿ" : "é—œé–‰";
      ui->status_label->setText(QString("ç‹€æ…‹: [F8] æ‰‹å‹•%1é»ƒç‡ˆ").arg(stateStr));
      blackbox->logEvent(QString("æ‰‹å‹•%1é»ƒç‡ˆ").arg(stateStr), 0);
    }
  });
}

void MainWindow::simulateAiTrigger(QString type) {
  // 1. æœ¬åœ°ç¡¬é«”é€£å‹• (é€é Blackbox é©…å‹•)
  if (type == "pig") {
    blackbox->setGpio(LED_RED, 1);
    blackbox->logEvent("AI æ¨¡æ“¬è§¸ç™¼: ç™¼ç¾å°è±¬å…¥ä¾µ (æœ€é«˜è­¦å ±)", 2);

    // åˆå§‹é³´å«
    if (!m_isMuted) {
      blackbox->setGpio(BUZZER, 1);
      QTimer::singleShot(200, [this]() {
        if (emergency->isBombActive())
          blackbox->setGpio(BUZZER, 0);
      });
    }

    // å•Ÿå‹• 5 åˆ†é˜ç‚¸å½ˆå€’æ•¸ (Kernel Timer)
    emergency->triggerPigBomb(5);
  } else if (type == "stranger") {
    blackbox->setGpio(LED_BLUE, 1);
    blackbox->logEvent("AI æ¨¡æ“¬è§¸ç™¼: ç™¼ç¾é™Œç”Ÿäºº", 1);
  }

  // è¨­å®š SecurityController é€²å…¥è­¦å ±é–å®šç‹€æ…‹
  QMetaObject::invokeMethod(security, "setAlarmActive", Q_ARG(bool, true));

  // 2. å¯«å…¥ JSON çµ¦ Web Server è®€å–
  QFile file("/tmp/guardian_alarm_status.json");
  if (file.open(QIODevice::WriteOnly | QIODevice::Text)) {
    QTextStream out(&file);
    out.setCodec("UTF-8"); // ç¢ºä¿ UTF-8 ç·¨ç¢¼
    out << "{\n"
        << "  \"alarm_active\": true,\n"
        << "  \"alarm_type\": \"" << type << "\",\n"
        << "  \"timestamp\": \""
        << QDateTime::currentDateTime().toString("yyyy-MM-dd HH:mm:ss")
        << "\",\n"
        << "  \"confidence\": 0.98\n"
        << "}";
    file.close();
    qDebug() << "è­¦å ±ç‹€æ…‹å·²å¯«å…¥ /tmp/guardian_alarm_status.json";
  }

  // 3. åŒæ­¥ç™¼é€ Discord æ¨æ’­
  sendDiscordNotification(type, (type == "pig" ? "high" : "normal"));
}

void MainWindow::sendDiscordNotification(QString type, QString priority) {
  QFile queueFile("/tmp/guardian_discord_queue.json");
  if (queueFile.open(QIODevice::WriteOnly | QIODevice::Text)) {
    QJsonObject jsonObj;
    jsonObj["type"] = (type == "pig" ? "pig_intrusion" : "stranger_detected");
    jsonObj["priority"] = priority;
    jsonObj["message"] =
        (type == "pig") ? QString::fromUtf8("ğŸš¨ åµæ¸¬åˆ°å°è±¬å…¥ä¾µï¼(æœ€é«˜è­¦å ±)")
                        : QString::fromUtf8("ğŸ‘¤ åµæ¸¬åˆ°é™Œç”Ÿäººä¾†è¨ª");
    jsonObj["timestamp"] =
        QDateTime::currentDateTime().toString("yyyy-MM-dd HH:mm:ss");
    jsonObj["image_path"] = "/tmp/guardian_images/alert.jpg";

    QJsonDocument doc(jsonObj);
    queueFile.write(doc.toJson());
    queueFile.close();
    qDebug() << "Discord æ¨æ’­ä»»å‹™å·²åŠ å…¥ä½‡åˆ—";
  } else {
    qDebug() << "ç„¡æ³•å¯«å…¥ Discord ä½‡åˆ—æª”";
  }
}

void MainWindow::sendDiscordCode(QString code) {
  QFile queueFile("/tmp/guardian_discord_queue.json");
  if (queueFile.open(QIODevice::WriteOnly | QIODevice::Text)) {
    QJsonObject jsonObj;
    jsonObj["type"] = "verification_code";
    jsonObj["priority"] = "high";
    jsonObj["message"] =
        QString::fromUtf8("æ‚¨çš„é ç«¯è§£é–é©—è­‰ç¢¼ç‚ºï¼š**") + code +
        QString::fromUtf8("**\nè«‹åœ¨ç¾å ´è¨­å‚™è¼¸å…¥æ­¤ä»£ç¢¼ä»¥å®Œæˆè§£é–ã€‚");
    jsonObj["timestamp"] =
        QDateTime::currentDateTime().toString("yyyy-MM-dd HH:mm:ss");

    QJsonDocument doc(jsonObj);
    queueFile.write(doc.toJson());
    queueFile.close();
    qDebug() << "é©—è­‰ç¢¼æ¨æ’­å·²åŠ å…¥ä½‡åˆ—:" << code;
  } else {
    qDebug() << "ç„¡æ³•å¯«å…¥ Discord ä½‡åˆ—æª” (é©—è­‰ç¢¼)";
  }
}

void MainWindow::updateAlarmJsonWithCountdown(int seconds, QString formatted) {
  QFile file("/tmp/guardian_alarm_status.json");
  // åªæœ‰åœ¨è­¦å ±æª”æ¡ˆå­˜åœ¨æ™‚æ‰æ›´æ–°ï¼ˆé¿å…èª¤å»ºç«‹ï¼‰
  if (file.exists() && file.open(QIODevice::ReadOnly | QIODevice::Text)) {
    QString content = file.readAll();
    file.close();

    // ç°¡å–®æ›¿æ›æˆ–è§£æã€‚é€™è£¡æ¡ç”¨ç°¡å–®æ›¿æ›é‚è¼¯ï¼Œç¢ºä¿ Web ç«¯èƒ½è®€åˆ°
    if (file.open(QIODevice::WriteOnly | QIODevice::Text)) {
      QTextStream out(&file);
      out.setCodec("UTF-8");

      // å¦‚æœå·²ç¶“æœ‰ countdown æ¬„ä½å‰‡æ›¿æ›ï¼Œæ²’æœ‰å‰‡æ’å…¥
      if (content.contains("\"countdown\"")) {
        // æ­£å‰‡è¡¨é”å¼æ›¿æ›æ¯”è¼ƒç©©å¥ï¼Œä½†é€™è£¡å…ˆç”¨ç°¡å–®é‚è¼¯
        // å¯¦éš›é–‹ç™¼å»ºè­°ä½¿ç”¨ QJsonDocument
      }

      // ç‚ºäº†å¿«é€Ÿæ¼”ç¤ºï¼Œæˆ‘å€‘ç›´æ¥é‡å¯«åŸºæœ¬çš„ JSON
      out << "{\n"
          << "  \"alarm_active\": true,\n"
          << "  \"alarm_type\": \"pig\",\n"
          << "  \"countdown\": " << seconds << ",\n"
          << "  \"countdown_str\": \"" << formatted << "\",\n"
          << "  \"timestamp\": \""
          << QDateTime::currentDateTime().toString("yyyy-MM-dd HH:mm:ss")
          << "\",\n"
          << "  \"confidence\": 0.98\n"
          << "}";
      file.close();
    }
  }
}

void MainWindow::handleShortcut(int keyId) {
  switch (keyId) {
  case 1:
    ui->status_label->setText("ç‹€æ…‹: [F1] é–‹é–€ä¸­(ç¶ è‰²LED äº®5ç§’)");
    blackbox->logEvent("é–‹é–€ä¸­(ç¶ è‰²LED äº®5ç§’)", 0);
    blackbox->setGpio(LED_GREEN, 1);
    QTimer::singleShot(5000, [this]() {
      blackbox->setGpio(LED_GREEN, 0);
      blackbox->logEvent("é—œé–€(ç¶ è‰²LED æš—)", 0);
    });
    break;
  case 2:
    m_isMuted = !m_isMuted; // åˆ‡æ›éœéŸ³ç‹€æ…‹
    if (m_isMuted) {
      ui->status_label->setText("ç‹€æ…‹: [F2] è­¦å ±å·²éœéŸ³");
      blackbox->logEvent("è­¦å ±éœéŸ³ (F2)", 1);
      blackbox->setGpio(LED_RED, 0);
      blackbox->setGpio(BUZZER, 0);
    } else {
      ui->status_label->setText("ç‹€æ…‹: [F2] è­¦å ±éŸ³æ•ˆå·²æ¢å¾©");
      blackbox->logEvent("æ¢å¾©è­¦å ±éŸ³æ•ˆ (F2)", 0);
      if (emergency->isBombActive()) {
        blackbox->setGpio(LED_RED, 1);
      }
    }
    break;
  case 3:
    ui->status_label->setText("ç‹€æ…‹: [F3] æŸ¥çœ‹æ—¥èªŒ");
    {
      QString logs = m_logHistory.join("\n");
      QMessageBox::information(this, "é»‘ç›’å­æ—¥èªŒ (æœ€è¿‘ 100 ç­†)",
                               logs.isEmpty() ? "ç„¡æ—¥èªŒ" : logs);
    }
    break;
  case 4:
    blackbox->logEvent("ç³»çµ±æ­£å¸¸é—œé–‰", 0);
    qApp->quit();
    break;
  }
}

void MainWindow::keyPressEvent(QKeyEvent *event) {
  if (event->key() == Qt::Key_F12) {
    if (m_f12Timer.isValid() && m_f12Timer.elapsed() < 500) {
      // 500ms å…§é€£æŒ‰å…©ä¸‹ F12
      qDebug() << "F12 é€£æŒ‰å…©ä¸‹ï¼šæ‰‹å‹•è§¸ç™¼ç·Šæ€¥è‡ªæ¯€ï¼";
      ui->status_label->setText(
          "<font color='red'><b>ğŸ’¥ F12 æ‰‹å‹•è§¸ç™¼è‡ªæ¯€ç¨‹åº ğŸ’¥</b></font>");
      blackbox->logEvent("F12 é€£æŒ‰å…©ä¸‹ï¼šæ‰‹å‹•è§¸ç™¼ç·Šæ€¥è‡ªæ¯€ç¨‹åº", 2);
      blackbox->setGpio(LED_RED, 1);
      emergency->triggerPigBomb(0);           // ç«‹å³è§¸ç™¼
      sendDiscordNotification("pig", "high"); // ç™¼é€ Discord é€šçŸ¥
      m_f12Timer.invalidate();
    } else {
      m_f12Timer.start();
    }
  }
  QMainWindow::keyPressEvent(event);
}

void MainWindow::updateFrame(QImage img) {
  if (!img.isNull()) {
    QSize labelSize = ui->video_label->size();
    if (labelSize.width() > 0 && labelSize.height() > 0) {
      ui->video_label->setPixmap(QPixmap::fromImage(img).scaled(
          labelSize, Qt::KeepAspectRatio, Qt::SmoothTransformation));
    } else {
      ui->video_label->setPixmap(QPixmap::fromImage(img));
    }
  }
}

void MainWindow::handlePasswordInput() {
  QString input = ui->password_input->text();
  QMetaObject::invokeMethod(security, "verifyPassword", Q_ARG(QString, input));
  ui->password_input->clear();
}
